improver_tests/utilities/test_InterpolateUsingDifference.py

```diff
@@ -0,0 +1,258 @@
+                               [np.nan, np.nan, np.nan],
+                               [15., 15., 15.]], dtype=np.float32)
+        sleet_rain = np.ma.masked_invalid(sleet_rain)
+
+        self.snow_sleet.data = snow_sleet
+        self.sleet_rain.data = sleet_rain
+
+        expected_unlimited = np.array([[5.0, 5.0, 5.0],
+                                       [8.5, 8.5, 8.5],
+                                       [15., 15., 15.]], dtype=np.float32)
+        expected_limited = np.array([[5.0, 5.0, 5.0],
+                                     [10., 10., 10.],
+                                     [15., 15., 15.]], dtype=np.float32)
+
+        result_unlimited = InterpolateUsingDifference().process(
+            self.sleet_rain, self.snow_sleet)
+
+        result_limited = InterpolateUsingDifference().process(
+            self.sleet_rain, self.snow_sleet, limit=self.snow_sleet,
+            limit_as_maximum=False)
+
+        assert_array_equal(result_unlimited.data, expected_unlimited)
+        assert_array_equal(result_limited.data, expected_limited)
+
+    def test_linear_failure(self):
+        """Test that if the use of linear interpolation does not result in a
+        complete difference field, and thus a complete field of interest, the
+        secondary use of nearest neighbour interpolation completes the
+        field."""
+
+        sleet_rain = np.array([[np.nan, np.nan, 4.0],
+                               [np.nan, np.nan, np.nan],
+                               [3.0, 3.0, 3.0]], dtype=np.float32)
+        sleet_rain = np.ma.masked_invalid(sleet_rain)
+        self.sleet_rain.data = sleet_rain
+
+        expected = np.array([[3.5, 4.0, 4.0],
+                             [8.5, 8.5, 8.5],
+                             [3.0, 3.0, 3.0]], dtype=np.float32)
+
+        result = InterpolateUsingDifference().process(self.sleet_rain,
+                                                      self.snow_sleet)
+
+        assert_array_equal(result.data, expected)
+        self.assertEqual(result.coords(), self.sleet_rain.coords())
+        self.assertEqual(result.metadata, self.sleet_rain.metadata)
+
+    @ManageWarnings(record=True)
+    def test_unmasked_input_cube(self, warning_list=None):
+        """Test a warning is raised if the input cube is not masked and that
+        the input cube is returned unchanged."""
+
+        self.sleet_rain.data = np.ones((3, 3), dtype=np.float32)
+        expected = self.sleet_rain.copy()
+        warning_msg = "Input cube unmasked, no data to fill in, returning"
+
+        result = InterpolateUsingDifference().process(
+                self.sleet_rain, self.snow_sleet)
+
+        self.assertEqual(result, expected)
+        self.assertTrue(any(item.category == UserWarning
+                            for item in warning_list))
+        self.assertTrue(any(warning_msg in str(item)
+                            for item in warning_list))
+
+    # Tests for the private _check_inputs method.
```

Comment:
Are you planning to move these into a separate class?

Reply:
Yes, it was all a rush on Friday in an attempt to get this reviewed whilst I was on leave. Moved.